### PR TITLE
Mutable N-dim arrays

### PR DESCRIPTION
**Pull request by Upabjojr**
Step 1 in porting PR https://github.com/sympy/sympy/pull/9112 into SymPy.

This PR adds various classes, all derived from _NDimArray_. The class structure is meant to resemble that of the matrices (sparse/dense, mutable/immutable), though immutable N-dim arrays are not implemented.

I have rewritten most of the code, as the indexing of dense matrices in https://github.com/sympy/sympy/pull/9112 was quite confusing and cumbersome.

### CODE REVIEW

**Comment by Upabjojr** [comment]
By the way, the current syntax to create a matrix from an iterable is:

```
Matrix(number_rows, number_colums, iterable)
```

I kept the parameter order to be compatible with the matrix module, so one may call `MutableDenseNDimArray(dim1, dim2, dim3, ... , iterable)`.

Though, I think it would be more confortable to postpone them to the _iterable_, so one may easily call:

```
MutableDenseNDimArray(iterable, *shape)
```
instead of
```
MutableDenseNDimArray(*(shape + (iterable,)))
```

**Comment by asmeurer** [comment]
You won't be able to use it in library code obviously, but in Python 3.5 you can write `MutableDenseNDimArray(*shape, iterable)`.

**Comment by Upabjojr** [comment]
@asmeurer the question was whether to invert the order or maintain it. That is, change the API and postpone the shape arguments after the iterable or not?

**Comment by asmeurer** [comment]
Another option would be to have the shape be a single argument, like `MutableDenseNDimArray(shape, iterable)` (then the order doesn't really matter).

**Comment by asmeurer** [comment]
I'm sure we'll want immutable versions eventually, as the mutable versions will be completely excluded being used by any other SymPy class. What is the advantage of having mutable versions at all?

**Comment by certik** [comment]
Looks ok in general. What is the difference to NumPy's multidimensional arrays?

**Comment by Upabjojr** [comment]
> What is the advantage of having mutable versions at all?

* It's easier to construct objects, exactly as in the Matrix module.
* Modifying an element does not need all other elements to be copied to a new data structure.
* Immutable types would have made this PR too large, maybe in another one.

> Looks ok in general. What is the difference to NumPy's multidimensional arrays?

* Every data you store gets sympified.
* Only mathematically defined operations should be implemented (no broadcasting and other stuff of numpy).
* no numpy dependency required.
* In the future, implement the immutable versions, which will be SymPy objects to use in expression directly.
* `sympy.diffgeom` could return one of these types or a type derived from these instead of a series of nested tuples/lists when deriving the Riemann tensor.

This PR is ready to me. Tell me if you spot anything you'd like to change.

**Comment by Upabjojr** [comment]
> Looks ok in general. What is the difference to NumPy's multidimensional arrays?

Also, SymPy specific methods could be implemented in the future, like derivatives, limits, etc..., it's about similar reasons to have SymPy matrices instead of using numpy ones.

**Comment by jksuom** [comment]
Looks good, though I haven't scanned everything (yet).

File: sympy/tensor/array/ndim_array.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        >>> from sympy.tensor.array.dense_ndim_array import MutableDenseNDimArray
+        >>> a = MutableDenseNDimArray.zeros(3,4,5,6,3)
+        >>> a.rank()
+        5
+
+        """
+        return self._rank
+
+    def __str__(self):
+        """Returns string, allows to use standart functions print() and str().
```
Comment (jksuom) [inline sympy/tensor/array/ndim_array.py]: standart -> standard

File: sympy/tensor/array/dense_ndim_array.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        >>> b = a.tomatrix()
+        >>> b
+        Matrix([
+        [1, 1, 1],
+        [1, 1, 1],
+        [1, 1, 1]])
+
+        """
+        if self.rank() != 2:
+            raise ValueError('Dimensions must be of size of 2')
```
Comment (jksuom) [inline sympy/tensor/array/dense_ndim_array.py]: Perhaps this could say more explicitly that a matrix is expected to be two-dimensional.

**Comment by jksuom** [comment]
I think this could go in. The number of commits should most probably be reduced. But before that is done I would like to hear some other comments.

**Comment by Upabjojr** [comment]
This PR has been rebased.

**Comment by certik** [comment]
@asmeurer what is your opinion on this PR? Should we merge it?

**Comment by certik** [comment]
@asmeurer, @mrocklin ping. Let me know your thoughts on this.

File: sympy/tensor/array/dense_ndim_array.py (hunk 2 above)
Comment (asmeurer) [inline sympy/tensor/array/dense_ndim_array.py]: I would also be more careful about the use of "rank" vs. "dimension".

**Comment by Upabjojr** [comment]
Sorry I messed up my branch on github. Refer to https://github.com/sympy/sympy/pull/10136/files

**Pull request closed without merging**